File: jafu/src/main/java/org/springframework/fu/jafu/redis/RedisDsl.java
package org.springframework.fu.jafu.redis;

import org.springframework.boot.autoconfigure.data.redis.JedisRedisInitializer;
import org.springframework.boot.autoconfigure.data.redis.RedisInitializer;
import org.springframework.boot.autoconfigure.data.redis.RedisProperties;
import org.springframework.context.ApplicationContextInitializer;
import org.springframework.context.support.GenericApplicationContext;

import java.util.function.Consumer;

/**
 * JaFu DSL for Redis configuration.
 *
 * @author Andreas Gebhardt
 */
public class RedisDsl extends AbstractRedisDsl<RedisDsl> {

    public RedisDsl(final Consumer<RedisDsl> dsl) {
        super(dsl);
    }

    public RedisDsl jedis() {
        return jedis(dsl -> {
        });
    }

    @Override
    protected RedisDsl getSelf() {
        return this;
    }

    public RedisDsl jedis(final Consumer<JedisDsl> jedis) {
        setRedisClientInitializer(new JedisRedisInitializer(properties));
        jedis.accept(new JedisDsl(properties.getJedis()));
        return this;
    }

    public static ApplicationContextInitializer<GenericApplicationContext> redis() {
        return new RedisDsl(dsl -> {
        });
    }

    public static ApplicationContextInitializer<GenericApplicationContext> redis(final Consumer<RedisDsl> dsl) {
        return new RedisDsl(dsl);
    }

    public static class JedisDsl {

        private final RedisProperties.Jedis jedis;

        JedisDsl(final RedisProperties.Jedis jedis) {
            this.jedis = jedis;
            this.jedis.setPool(new RedisProperties.Pool());
        }

        public JedisDsl pool(final Consumer<PoolDsl> pool) {
            pool.accept(new PoolDsl(this.jedis.getPool()));
            return this;
        }

    }

    @Override
    public void initialize(GenericApplicationContext context) {
        super.initialize(context);
        new RedisInitializer().initialize(context);
    }

}

File: kofu/src/main/kotlin/org/springframework/fu/kofu/redis/RedisDsl.kt
package org.springframework.fu.kofu.redis

import org.springframework.boot.autoconfigure.data.redis.ClusterInitializer
import org.springframework.boot.autoconfigure.data.redis.JedisRedisInitializer
import org.springframework.boot.autoconfigure.data.redis.LettuceRedisInitializer
import org.springframework.boot.autoconfigure.data.redis.RedisInitializer
import org.springframework.boot.autoconfigure.data.redis.RedisProperties.Pool
import org.springframework.boot.autoconfigure.data.redis.SentinelInitializer
import org.springframework.context.ApplicationContextInitializer
import org.springframework.context.support.GenericApplicationContext
import org.springframework.fu.kofu.ConfigurationDsl

/**
 * Kofu DSL for Redis configuration
 *
 * @author Waldemar Panas
 * @author Sebastien Deleuze
 */
@Suppress("UsePropertyAccessSyntax")
class RedisDsl(private val init: RedisDsl.() -> Unit) : AbstractRedisDsl(), JedisRedisSupporter, LettuceRedisSupporter {

	private var jedisInitializer: ApplicationContextInitializer<GenericApplicationContext>? = null

	private var lettuceInitializer: ApplicationContextInitializer<GenericApplicationContext>? = null


	override fun jedis(dsl: PoolDsl.() -> Unit) {
		jedisInitializer = ApplicationContextInitializer {
			JedisDsl(properties, dsl).initialize(it)
			JedisRedisInitializer(properties).initialize(it)
		}
	}

	override fun lettuce(dsl: LettuceDsl.() -> Unit) {
		lettuceInitializer = ApplicationContextInitializer {
			LettuceDsl(properties, dsl).initialize(context)
			LettuceRedisInitializer(properties).initialize(context)
		}
	}

	override fun initialize(context: GenericApplicationContext) {
		super.initialize(context)
		init()
		if (jedisInitializer != null) {
			jedisInitializer!!.initialize(context)
		}
		else {
			if (lettuceInitializer == null) lettuce()
			lettuceInitializer!!.initialize(context)
		}
		RedisInitializer().initialize(context)
		ClusterInitializer(properties.cluster).initialize(context)
		SentinelInitializer(properties.sentinel).initialize(context)
	}
}


/**
 * Configure Redis support by registering 2 beans: `redisTemplate` of type `RedisTemplate<Any, Any>`
 * and `stringRedisTemplate` of type `StringRedisTemplate`. `jedis()` or `lettuce()` can be used
 * to select the driver, Lettuce will be used by default if none is used.
 * @see RedisDsl
 */
fun ConfigurationDsl.redis(dsl: RedisDsl.() -> Unit = {}) {
	RedisDsl(dsl).initialize(context)
}
